File: benchmarks/bench_roundtrip.cpp
// bench_roundtrip.cpp — localhost WebSocket echo latency (wscpp)

#include "bench_common.hpp"
#include <wscpp/client.hpp>
#include <wscpp/server.hpp>
#include <asio/io_context.hpp>
#include <asio/ip/tcp.hpp>
#include <atomic>
#include <cstdio>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

using namespace wscpp;
using namespace wscpp::bench;

namespace {

uint16_t pick_free_port() {
    asio::io_context io;
    asio::ip::tcp::acceptor acc(io, asio::ip::tcp::endpoint(asio::ip::tcp::v4(), 0));
    return acc.local_endpoint().port();
}

} // namespace

int main() {
    const int samples = 100;
    const uint16_t port = pick_free_port();

    server srv;
    srv.set_on_connection([&](std::shared_ptr<connection> conn) {
        conn->set_on_message([conn](const std::vector<uint8_t>& data, frame::opcode) {
            conn->send_text(std::string(data.begin(), data.end()));
        });
    });
    srv.listen(port);
    srv.start();

    std::vector<double> latencies_ms;
    latencies_ms.reserve(static_cast<std::size_t>(samples));
    std::mutex latency_mutex;
    std::vector<clock::time_point> send_times(static_cast<std::size_t>(samples));
    std::atomic<bool> done(false);

    std::thread client_thread([&]() {
        client cli;
        int next_send = 0;

        cli.set_on_open([&]() {
            send_times[0] = clock::now();
            cli.send_text("ping-0");
            next_send = 1;
        });

        cli.set_on_message([&](const std::vector<uint8_t>&, frame::opcode) {
            const int n = next_send - 1;
            if (n >= 0 && n < samples) {
                const double ms = elapsed_sec(send_times[static_cast<std::size_t>(n)], clock::now()) * 1000.0;
                {
                    std::lock_guard<std::mutex> lock(latency_mutex);
                    latencies_ms.push_back(ms);
                }
            }
            if (next_send < samples) {
                send_times[static_cast<std::size_t>(next_send)] = clock::now();
                cli.send_text("ping-" + std::to_string(next_send));
                ++next_send;
            } else {
                cli.close();
                done = true;
            }
        });

        try {
            cli.connect("ws://127.0.0.1:" + std::to_string(port) + "/");
            while (!done.load()) {
                std::this_thread::sleep_for(std::chrono::milliseconds(1));
            }
        } catch (...) {
            done = true;
        }
    });

    while (!done.load()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
    client_thread.join();
    srv.stop();

    std::printf("wscpp roundtrip (%zu samples, port %u)\n", latencies_ms.size(), port);
    print_latency("echo_latency", latencies_ms);

    const std::size_t throughput_bytes = 64 * 1024;
    std::vector<uint8_t> blob(throughput_bytes, 0xAB);
    std::atomic<bool> tp_done(false);
    int tp_iterations = 0;

    const uint16_t tp_port = pick_free_port();
    server tp_srv;
    tp_srv.set_on_connection([&](std::shared_ptr<connection> conn) {
        conn->set_on_message([conn](const std::vector<uint8_t>& data, frame::opcode op) {
            if (op == frame::opcode::BINARY) {
                conn->send_binary(data.data(), data.size());
            }
        });
    });
    tp_srv.listen(tp_port);
    tp_srv.start();

    std::thread tp_thread([&]() {
        client cli;
        int sent = 0;
        cli.set_on_open([&]() { cli.send_binary(blob.data(), blob.size()); });
        cli.set_on_message([&](const std::vector<uint8_t>& data, frame::opcode op) {
            if (op != frame::opcode::BINARY || data.size() != throughput_bytes) {
                return;
            }
            ++sent;
            if (sent < 100) {
                cli.send_binary(blob.data(), blob.size());
            } else {
                tp_iterations = sent;
                cli.close();
                tp_done = true;
            }
        });
        try {
            cli.connect("ws://127.0.0.1:" + std::to_string(tp_port) + "/");
            while (!tp_done.load()) {
                std::this_thread::sleep_for(std::chrono::milliseconds(1));
            }
        } catch (...) {
            tp_done = true;
        }
    });

    const clock::time_point tp_start = clock::now();
    while (!tp_done.load()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
    const clock::time_point tp_end = clock::now();
    tp_thread.join();
    tp_srv.stop();

    print_throughput("echo_throughput_64k",
                     static_cast<double>(throughput_bytes * tp_iterations),
                     elapsed_sec(tp_start, tp_end), tp_iterations);

    return 0;
}

File: rust/ws-rs-benches/src/bin/bench_roundtrip.rs
// bench_roundtrip — localhost WebSocket echo latency + 64 KiB throughput (mirrors C++ bench_roundtrip)

use std::time::Instant;

use ws_rs::client::Client;
use ws_rs_benches::common::{
    pick_free_port, print_banner, print_latency, print_throughput, sleep_ms, spawn_echo_server,
};

const SAMPLES: usize = 100;
const THROUGHPUT_BYTES: usize = 64 * 1024;
const THROUGHPUT_ITERATIONS: i32 = 100;

#[tokio::main]
async fn main() {
    print_banner("bench_roundtrip");

    let port = pick_free_port().await;
    let _server = spawn_echo_server("127.0.0.1", port).await;
    sleep_ms(50).await;

    let url = format!("ws://127.0.0.1:{port}/");
    let mut client = Client::new();
    if client.connect(&url).await.is_err() {
        eprintln!("connect failed");
        std::process::exit(1);
    }

    let mut latencies_ms = Vec::with_capacity(SAMPLES);
    for i in 0..SAMPLES {
        let msg = format!("ping-{i}");
        let start = Instant::now();
        if client.send_text(&msg).await.is_err() {
            break;
        }
        if client.recv_text().await.is_err() {
            break;
        }
        latencies_ms.push(ws_rs_benches::common::elapsed_sec(start, Instant::now()) * 1000.0);
    }
    client.close().await.ok();

    println!(
        "ws-rs roundtrip ({} samples, port {port})",
        latencies_ms.len()
    );
    print_latency("echo_latency", &latencies_ms);

    let tp_port = pick_free_port().await;
    let _tp_server = spawn_echo_server("127.0.0.1", tp_port).await;
    sleep_ms(50).await;

    let tp_url = format!("ws://127.0.0.1:{tp_port}/");
    let blob = vec![0xABu8; THROUGHPUT_BYTES];

    let tp_start = Instant::now();
    let mut tp_client = Client::new();
    if tp_client.connect(&tp_url).await.is_err() {
        eprintln!("throughput connect failed");
        std::process::exit(1);
    }

    let mut tp_iterations = 0i32;
    if tp_client.send_binary(&blob).await.is_ok() {
        for _ in 0..THROUGHPUT_ITERATIONS {
            match tp_client.recv_binary().await {
                Ok(Some(data)) if data.len() == THROUGHPUT_BYTES => {
                    tp_iterations += 1;
                    if tp_iterations < THROUGHPUT_ITERATIONS {
                        if tp_client.send_binary(&blob).await.is_err() {
                            break;
                        }
                    } else {
                        tp_client.close().await.ok();
                        break;
                    }
                }
                _ => break,
            }
        }
    }
    let tp_end = Instant::now();

    print_throughput(
        "echo_throughput_64k",
        THROUGHPUT_BYTES as f64 * tp_iterations as f64,
        ws_rs_benches::common::elapsed_sec(tp_start, tp_end),
        tp_iterations,
    );

    if latencies_ms.len() != SAMPLES {
        std::process::exit(1);
    }
}
